Exception handling in do_set

logstash crashes when memcached servers becomes unavailable. 

```
W, [2019-11-14T08:15:41.117628 #1]  WARN -- : memcached-03:11211 failed (count: 0) EOFError: End of file reached
W, [2019-11-14T08:15:41.320959 #1]  WARN -- : memcached-03:11211 failed (count: 0) Errno::EPIPE: Broken pipe - No message available
W, [2019-11-14T08:15:41.352324 #1]  WARN -- : memcached-03:11211 failed (count: 1) Errno::ECONNREFUSED: Connection refused - connect(2) for "memcached-03" port 11211
W, [2019-11-14T08:15:41.352323 #1]  WARN -- : memcached-03:11211 failed (count: 1) Errno::ECONNREFUSED: Connection refused - connect(2) for "memcached-03" port 11211
W, [2019-11-14T08:15:41.358390 #1]  WARN -- : memcached-03:11211 is down
W, [2019-11-14T08:15:41.359376 #1]  WARN -- : memcached-03:11211 is down
W, [2019-11-14T08:16:05.118757 #1]  WARN -- : memcached-02:11211 failed (count: 0) EOFError: End of file reached
W, [2019-11-14T08:16:05.151353 #1]  WARN -- : memcached-02:11211 failed (count: 0) Errno::EPIPE: Broken pipe - No message available
W, [2019-11-14T08:16:05.152609 #1]  WARN -- : memcached-02:11211 failed (count: 1) Errno::ECONNREFUSED: Connection refused - connect(2) for "memcached-02" port 11211
W, [2019-11-14T08:16:05.153428 #1]  WARN -- : memcached-02:11211 is down
W, [2019-11-14T08:16:05.165181 #1]  WARN -- : memcached-02:11211 failed (count: 1) Errno::ECONNREFUSED: Connection refused - connect(2) for "memcached-02" port 11211
W, [2019-11-14T08:16:05.165753 #1]  WARN -- : memcached-02:11211 is down
[2019-11-14T08:16:09,033][ERROR][org.logstash.execution.WorkerLoop] Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash.
org.jruby.exceptions.RuntimeError: (RingError) No server available
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.dalli_minus_2_dot_7_dot_10.lib.dalli.ring.server_for_key(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/dalli-2.7.10/lib/dalli/ring.rb:46) ~[?:?]
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.dalli_minus_2_dot_7_dot_10.lib.dalli.client.perform(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/dalli-2.7.10/lib/dalli/client.rb:367) ~[?:?]
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.dalli_minus_2_dot_7_dot_10.lib.dalli.client.set(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/dalli-2.7.10/lib/dalli/client.rb:130) ~[?:?]
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_filter_minus_memcached_minus_1_dot_0_dot_1.lib.logstash.filters.memcached.do_set(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-filter-memcached-1.0.1/lib/logstash/filters/memcached.rb:137) ~[?:?]
	at org.jruby.RubyHash.each(org/jruby/RubyHash.java:1419) ~[jruby-complete-9.2.7.0.jar:?]
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_filter_minus_memcached_minus_1_dot_0_dot_1.lib.logstash.filters.memcached.do_set(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-filter-memcached-1.0.1/lib/logstash/filters/memcached.rb:135) ~[?:?]
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.dalli_minus_2_dot_7_dot_10.lib.dalli.client.multi(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/dalli-2.7.10/lib/dalli/client.rb:50) ~[?:?]
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_filter_minus_memcached_minus_1_dot_0_dot_1.lib.logstash.filters.memcached.do_set(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-filter-memcached-1.0.1/lib/logstash/filters/memcached.rb:134) ~[?:?]
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_filter_minus_memcached_minus_1_dot_0_dot_1.lib.logstash.filters.memcached.filter(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-filter-memcached-1.0.1/lib/logstash/filters/memcached.rb:84) ~[?:?]
	at usr.share.logstash.logstash_minus_core.lib.logstash.filters.base.do_filter(/usr/share/logstash/logstash-core/lib/logstash/filters/base.rb:143) ~[?:?]
	at usr.share.logstash.logstash_minus_core.lib.logstash.filters.base.multi_filter(/usr/share/logstash/logstash-core/lib/logstash/filters/base.rb:162) ~[?:?]
	at org.jruby.RubyArray.each(org/jruby/RubyArray.java:1792) ~[jruby-complete-9.2.7.0.jar:?]
	at usr.share.logstash.logstash_minus_core.lib.logstash.filters.base.multi_filter(/usr/share/logstash/logstash-core/lib/logstash/filters/base.rb:159) ~[?:?]
	at org.logstash.config.ir.compiler.AbstractFilterDelegatorExt.multi_filter(org/logstash/config/ir/compiler/AbstractFilterDelegatorExt.java:115) ~[logstash-core.jar:?]
	at usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.start_workers(/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:239) ~[?:?]
warning: thread "[p049_arcsight-asa-01-flow]>worker11" terminated with exception (report_on_exception is true):
java.lang.IllegalStateException: org.jruby.exceptions.RuntimeError: (RingError) No server available
	at org.logstash.execution.WorkerLoop.run(org/logstash/execution/WorkerLoop.java:85)
	at jdk.internal.reflect.GeneratedMethodAccessor82.invoke(jdk/internal/reflect/GeneratedMethodAccessor82)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(jdk/internal/reflect/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:566)
	at org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:440)
	at org.jruby.javasupport.JavaMethod.invokeDirect(org/jruby/javasupport/JavaMethod.java:304)
	at usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.start_workers(/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:239)
	at org.jruby.RubyProc.call(org/jruby/RubyProc.java:295)
	at org.jruby.RubyProc.call(org/jruby/RubyProc.java:274)
	at org.jruby.RubyProc.call(org/jruby/RubyProc.java:270)
	at java.lang.Thread.run(java/lang/Thread.java:834)
Caused by: org.jruby.exceptions.RuntimeError: (RingError) No server available
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.dalli_minus_2_dot_7_dot_10.lib.dalli.ring.server_for_key(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/dalli-2.7.10/lib/dalli/ring.rb:46)
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.dalli_minus_2_dot_7_dot_10.lib.dalli.client.perform(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/dalli-2.7.10/lib/dalli/client.rb:367)
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.dalli_minus_2_dot_7_dot_10.lib.dalli.client.set(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/dalli-2.7.10/lib/dalli/client.rb:130)
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_filter_minus_memcached_minus_1_dot_0_dot_1.lib.logstash.filters.memcached.do_set(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-filter-memcached-1.0.1/lib/logstash/filters/memcached.rb:137)
	at org.jruby.RubyHash.each(org/jruby/RubyHash.java:1419)
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_filter_minus_memcached_minus_1_dot_0_dot_1.lib.logstash.filters.memcached.do_set(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-filter-memcached-1.0.1/lib/logstash/filters/memcached.rb:135)
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.dalli_minus_2_dot_7_dot_10.lib.dalli.client.multi(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/dalli-2.7.10/lib/dalli/client.rb:50)
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_filter_minus_memcached_minus_1_dot_0_dot_1.lib.logstash.filters.memcached.do_set(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-filter-memcached-1.0.1/lib/logstash/filters/memcached.rb:134)
	at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_filter_minus_memcached_minus_1_dot_0_dot_1.lib.logstash.filters.memcached.filter(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-filter-memcached-1.0.1/lib/logstash/filters/memcached.rb:84)
	at usr.share.logstash.logstash_minus_core.lib.logstash.filters.base.do_filter(/usr/share/logstash/logstash-core/lib/logstash/filters/base.rb:143)
	at usr.share.logstash.logstash_minus_core.lib.logstash.filters.base.multi_filter(/usr/share/logstash/logstash-core/lib/logstash/filters/base.rb:162)
	at org.jruby.RubyArray.each(org/jruby/RubyArray.java:1792)
	at usr.share.logstash.logstash_minus_core.lib.logstash.filters.base.multi_filter(/usr/share/logstash/logstash-core/lib/logstash/filters/base.rb:159)
	at org.logstash.config.ir.compiler.AbstractFilterDelegatorExt.multi_filter(org/logstash/config/ir/compiler/AbstractFilterDelegatorExt.java:115)
	at usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.start_workers(/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:239)
[2019-11-14T08:16:09,153][FATAL][logstash.runner          ] An unexpected error occurred! {:error=>java.lang.IllegalStateException: org.jruby.exceptions.RuntimeError: (RingError) No server available, :backtrace=>["org.logstash.execution.WorkerLoop.run(org/logstash/execution/WorkerLoop.java:85)", "jdk.internal.reflect.GeneratedMethodAccessor82.invoke(jdk/internal/reflect/GeneratedMethodAccessor82)", "jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(jdk/internal/reflect/DelegatingMethodAccessorImpl.java:43)", "java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:566)", "org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:440)", "org.jruby.javasupport.JavaMethod.invokeDirect(org/jruby/javasupport/JavaMethod.java:304)", "usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.start_workers(/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:239)", "org.jruby.RubyProc.call(org/jruby/RubyProc.java:295)", "org.jruby.RubyProc.call(org/jruby/RubyProc.java:274)", "org.jruby.RubyProc.call(org/jruby/RubyProc.java:270)", "java.lang.Thread.run(java/lang/Thread.java:834)"]}
[2019-11-14T08:16:09,317][ERROR][org.logstash.Logstash    ] java.lang.IllegalStateException: Logstash stopped processing because of an error: (SystemExit) exit
[2019-11-14T08:16:09,477][ERROR][logstash.inputs.udp      ] UDP listener died {:exception=>#<IOError: closed stream>, :backtrace=>["org/jruby/RubyIO.java:3552:in `select'", ":1:in `'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-input-udp-3.3.4/lib/logstash/inputs/udp.rb:68:in `run'", "/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:309:in `inputworker'", "/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:302:in `block in start_input'"]}
```

It appears the dalli client eventually throws a `RuntimeError: (RingError) No server available` when no server is available. This appens at https://github.com/logstash-plugins/logstash-filter-memcached/blob/db53ae6d03a16338ca0a54cae406da7b16fe2b14/lib/logstash/filters/memcached.rb#L137 and there is no exception handling to handle such a condition which leads to crashing the plugin and logstash. 

it looks like we could improve this by catching this exception and probably go into a reconnection strategy?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,6 @@
+## 1.1.0
+ - Added better exception handling [#25](https://github.com/logstash-plugins/logstash-filter-memcached/pull/25)
+
 ## 1.0.2
  - Fixed issue with ttl not being set [#13](https://github.com/logstash-plugins/logstash-filter-memcached/pull/13)
 

--- a/CONTRIBUTORS
+++ b/CONTRIBUTORS
@@ -2,7 +2,8 @@ The following is a list of people who have contributed ideas, code, bug
 reports, or in general have helped logstash along its way.
 
 Contributors:
-* Ry Biesemeyer - ry.biesemeyer@elastic.co
+* Ry Biesemeyer (yaauie)
+* Colin Surprenant (colinsurprenant)
 
 Note: If you've sent us patches, bug reports, or otherwise contributed to
 Logstash, and you aren't on the list above and want to be, please let us know

--- a/docs/index.asciidoc
+++ b/docs/index.asciidoc
@@ -76,6 +76,7 @@ This plugin supports the following configuration options plus the <<plugins-{typ
 | <<plugins-{type}s-{plugin}-namespace>> |<<string,string>>|No
 | <<plugins-{type}s-{plugin}-get>> |<<hash,hash>>|No
 | <<plugins-{type}s-{plugin}-set>> |<<hash,hash>>|No
+| <<plugins-{type}s-{plugin}-tag_on_failure>> |<<string,string>>|No
 | <<plugins-{type}s-{plugin}-ttl>> |<<number,number>>|No
 |=======================================================================
 
@@ -163,6 +164,16 @@ filter {
   }
 }
 ----
+
+[id="plugins-{type}s-{plugin}-tag_on_failure"]
+===== `tag_on_failure`
+
+  * Value type is <<string,string>>
+  * The default value for this setting is `_memcached_failure`.
+
+When a memcached operation causes a runtime exception to be thrown within the plugin,
+the operation is safely aborted without crashing the plugin, and the event is
+tagged with the provided value.
 
 [id="plugins-{type}s-{plugin}-ttl"]
 ===== `ttl`

--- a/lib/logstash/filters/memcached.rb
+++ b/lib/logstash/filters/memcached.rb
@@ -3,7 +3,6 @@ require "logstash/filters/base"
 require "logstash/namespace"
 require 'dalli'
 
-
 # This filter provides facilities to interact with Memcached.
 class LogStash::Filters::Memcached < LogStash::Filters::Base
 

--- a/lib/logstash/filters/memcached.rb
+++ b/lib/logstash/filters/memcached.rb
@@ -1,6 +1,8 @@
 # encoding: utf-8
 require "logstash/filters/base"
 require "logstash/namespace"
+require 'dalli'
+
 
 # This filter provides facilities to interact with Memcached.
 class LogStash::Filters::Memcached < LogStash::Filters::Base
@@ -67,28 +69,59 @@ class LogStash::Filters::Memcached < LogStash::Filters::Base
   # NOTE: in Memcached, a value of 0 (default) means "never expire"
   config :ttl, :validate => :number, :default => 0
 
+  # Tags the event on failure. This can be used in later analysis.
+  config :tag_on_failure, :validate => :string, :default => "_memcached_failure"
+
   public
 
   attr_reader :cache
 
   def register
-    if @ttl < 0
-      logger.error("ttl cannot be negative")
-      fail("invalid ttl: cannot be negative")
-    end
+    raise(LogStash::ConfigurationError, "'ttl' option cannot be negative") if @ttl < 0
 
-    @cache = establish_connection
-  end # def register
+    @memcached_hosts = validate_connection_hosts
+    @memcached_options = validate_connection_options
+    begin
+      @cache = new_connection(@memcached_hosts, @memcached_options)
+    rescue => e
+      logger.error("failed to connect to memcached", hosts: @memcached_hosts, options: @memcached_options, message: e.message)
+      fail("failed to connect to memcached")
+    end
+    @connected = Concurrent::AtomicBoolean.new(true)
+    @connection_mutex = Mutex.new
+  end
 
   def filter(event)
-    set_success = do_set(event)
-    get_success = do_get(event)
+    unless connection_available?
+      event.tag(@tag_on_failure)
+      return
+    end
 
-    filter_matched(event) if (set_success || get_success)
-  end # def filter
+    begin
+      set_success = do_set(event)
+      get_success = do_get(event)
+      filter_matched(event) if (set_success || get_success)
+    rescue Dalli::NetworkError, Dalli::RingError => e
+      event.tag(@tag_on_failure)
+      logger.error("memcached communication error",  hosts: @memcached_hosts, options: @memcached_options, message: e.message)
+      close
+    rescue => e
+      meta = { :message => e.message }
+      meta[:backtrace] = e.backtrace if logger.debug?
+      logger.error("unexpected error", meta)
+      event.tag(@tag_on_failure)
+    end
+  end
 
   def close
-    cache.close
+    @connection_mutex.synchronize do
+      @connected.make_false
+      cache.close
+    end
+  rescue => e
+    # we basically ignore any error here as we may be trying to close an invalid
+    # connection or if we close on shutdown we can also ignore any error
+    logger.debug("error closing memcached connection", :message => e.message)
   end
 
   private
@@ -141,19 +174,33 @@ class LogStash::Filters::Memcached < LogStash::Filters::Base
     return true
   end
 
-  def establish_connection
-    require 'dalli'
-
-    hosts = validate_connection_hosts
-    options = validate_connection_options
+  def new_connection(hosts, options)
     logger.debug('connecting to memcached', context(hosts: hosts, options: options))
-    Dalli::Client.new(@hosts, options).tap do |client|
-      begin
-        client.alive!
-      rescue Dalli::RingError
-        logger.error("failed to connect", context(hosts: hosts, options: options))
-        fail("cannot connect to memcached")
-      end
+    Dalli::Client.new(hosts, options).tap { |client| client.alive! }
+  end
+
+  # reconnect is not thread safe
+  def reconnect(hosts, options)
+    begin
+      @cache = new_connection(hosts, options)
+      @connected.make_true
+    rescue => e
+      logger.error("failed to reconnect to memcached", hosts: hosts, options: options, message: e.message)
+      @connected.make_false
+    end
+    return @connected.value
+  end
+
+  def connection_available?
+    # this method is called at every #filter method invocation and to minimize synchronization cost
+    # only @connected if fist check. The tradeoff is that another worker connection could be in the
+    # process of failing and @connected will not yet reflect that but this is acceptable for performance reason.
+    return true if @connected.true?
+
+    @connection_mutex.synchronize do
+      # the reconnection process is exclusive and will not be be concurrently performed in another worker
+      # by re-verifying the state of @connected from the exclusive code.
+      return @connected.true? ? true : reconnect(@memcached_hosts, @memcached_options)
     end
   end
 
@@ -165,8 +212,7 @@ class LogStash::Filters::Memcached < LogStash::Filters::Base
   end
 
   def validate_connection_hosts
-    logger.error("configuration: hosts empty!") && fail if @hosts.empty?
-
+    raise(LogStash::ConfigurationError, "'hosts' cannot be empty") if @hosts.empty?
     @hosts.map(&:to_s)
   end
 
@@ -178,4 +224,4 @@ class LogStash::Filters::Memcached < LogStash::Filters::Base
 
     @plugin_context.merge(hash)
   end
-end # class LogStash::Filters::Memcached
+end

--- a/logstash-filter-memcached.gemspec
+++ b/logstash-filter-memcached.gemspec
@@ -3,9 +3,9 @@ Gem::Specification.new do |s|
   s.version       = '1.0.2'
   s.licenses      = ['Apache-2.0']
   s.summary       = 'A Logstash filter plugin for interacting with memcached'
-  s.homepage      = 'https://github.com/yaauie/logstash-filter-memcached'
-  s.authors       = ['Ry Biesemeyer']
-  s.email         = 'ry.biesemeyer@elastic.co'
+  s.authors       = ["Elastic"]
+  s.email         = 'info@elastic.co'
+  s.homepage      = "http://www.elastic.co/guide/en/logstash/current/index.html"
   s.require_paths = ['lib']
 
   # Files

--- a/logstash-filter-memcached.gemspec
+++ b/logstash-filter-memcached.gemspec
@@ -1,6 +1,6 @@
 Gem::Specification.new do |s|
   s.name          = 'logstash-filter-memcached'
-  s.version       = '1.0.2'
+  s.version       = '1.1.0'
   s.licenses      = ['Apache-2.0']
   s.summary       = 'A Logstash filter plugin for interacting with memcached'
   s.authors       = ["Elastic"]

--- a/spec/filters/memcached_spec.rb
+++ b/spec/filters/memcached_spec.rb
@@ -7,87 +7,150 @@ LogStash::Logging::Logger::configure_logging("TRACE")
 describe LogStash::Filters::Memcached do
   subject(:memcached_filter) { described_class.new(config) }
   let(:cache) { double('memcached') }
-  before(:each) do
-    allow(memcached_filter).to receive(:establish_connection).and_return(cache)
-    allow(memcached_filter).to receive(:close)
-    memcached_filter.register
+
+  describe "a well behaved memcached server" do
+    before(:each) do
+      allow(memcached_filter).to receive(:new_connection).and_return(cache)
+      allow(memcached_filter).to receive(:close)
+      memcached_filter.register
+    end
+
+    after(:each) do
+      memcached_filter.close
+    end
+
+    describe "#get" do
+      let(:event) { ::LogStash::Event.new(data) }
+      let(:config) do
+        {
+          "hosts" => ["localhost:11211"],
+          "get" => { "success/%{success}/answer" => "[ultimate][answer]" }
+        }
+      end
+      let(:data) { { "success" => "true" } }
+
+      it "retrieves data from memcache" do
+        expect(cache).to receive(:get_multi).with(["success/true/answer"]).and_return({"success/true/answer" => "42"})
+        expect(memcached_filter).to receive(:filter_matched)
+        memcached_filter.filter(event)
+        expect(event.get("ultimate")).to eq("answer" => "42")
+      end
+
+      context 'when memcached does not hold the value' do
+        before do
+          expect(cache).to receive(:get_multi).with(["success/true/answer"]).and_return({"success/true/answer" => nil})
+        end
+
+        it 'does not invoke `filter_matched`' do
+          expect(memcached_filter).to_not receive(:filter_matched)
+          memcached_filter.filter(event)
+        end
+
+        it 'does not populate the value' do
+          memcached_filter.filter(event)
+          expect(event.include?("ultimate")).to be false
+        end
+      end
+    end
+
+    describe "#set" do
+      let(:event) { ::LogStash::Event.new(data) }
+      let(:config) do
+        {
+          "hosts" => ["localhost:11211"],
+          "set" => { "[answer]" => "success/%{success}/answer" },
+        }
+      end
+      before(:each) do
+        allow(cache).to receive(:multi) {|&b| b.call }
+        allow(cache).to receive(:set)
+      end
+
+      context 'when the event includes the value to set' do
+        let(:data) { { "answer" => "42", "success" => "true" } }
+
+        it "sets data on memcached" do
+          expect(cache).to receive(:set).with("success/true/answer", "42")
+          memcached_filter.filter(event)
+        end
+
+        it 'invokes `filter_matched`' do
+          expect(memcached_filter).to receive(:filter_matched)
+          memcached_filter.filter(event)
+        end
+      end
+
+      context 'when the event does not include the value being set' do
+        let(:data) { { "success" => "true" } }
+
+        it 'does not set data on memcached' do
+          expect(cache).to_not receive(:set)
+          memcached_filter.filter(event)
+        end
+
+        it 'does not invoke `filter_matched`' do
+          expect(memcached_filter).to_not receive(:filter_matched)
+          memcached_filter.filter(event)
+        end
+      end
+    end
   end
 
-  after(:each) do
-    memcached_filter.close
-  end
-
-  describe "#get" do
-    let(:event) { ::LogStash::Event.new(data) }
+  describe "a misbehaved memcached server" do
     let(:config) do
       {
         "hosts" => ["localhost:11211"],
-        "get" => { "success/%{success}/answer" => "[ultimate][answer]" }
-      }
-    end
-    let(:data) { { "success" => "true" } }
-
-    it "retrieves data from memcache" do
-      expect(cache).to receive(:get_multi).with(["success/true/answer"]).and_return({"success/true/answer" => "42"})
-      expect(memcached_filter).to receive(:filter_matched)
-      memcached_filter.filter(event)
-      expect(event.get("ultimate")).to eq("answer" => "42")
-    end
-
-    context 'when memcached does not hold the value' do
-      before do
-        expect(cache).to receive(:get_multi).with(["success/true/answer"]).and_return({"success/true/answer" => nil})
-      end
-
-      it 'does not invoke `filter_matched`' do
-        expect(memcached_filter).to_not receive(:filter_matched)
-        memcached_filter.filter(event)
-      end
-
-      it 'does not populate the value' do
-        memcached_filter.filter(event)
-        expect(event.include?("ultimate")).to be false
-      end
-    end
-  end
-
-  describe "#set" do
-    let(:event) { ::LogStash::Event.new(data) }
-    let(:config) do
-      {
-        "hosts" => ["localhost:11211"],
+        "get" => { "success/%{success}/answer" => "[ultimate][answer]" },
         "set" => { "[answer]" => "success/%{success}/answer" },
       }
     end
-    before(:each) do
-      allow(cache).to receive(:multi) {|&b| b.call }
-      allow(cache).to receive(:set)
+
+    context "#register" do
+      it "raises error upon new connection error" do
+        allow(memcached_filter).to receive(:new_connection).and_raise("some error")
+        expect{memcached_filter.register}.to raise_error(RuntimeError)
+      end
     end
 
-    context 'when the event includes the value to set' do
+    context "#get and #set" do
+      let(:event) { ::LogStash::Event.new(data) }
       let(:data) { { "answer" => "42", "success" => "true" } }
 
-      it "sets data on memcached" do
-        expect(cache).to receive(:set).with("success/true/answer", "42")
-        memcached_filter.filter(event)
+      before(:each) do
+        allow(memcached_filter).to receive(:new_connection).and_return(cache)
+        allow(cache).to receive(:close).and_raise("some error")
+
+        memcached_filter.register
       end
 
-      it 'invokes `filter_matched`' do
-        expect(memcached_filter).to receive(:filter_matched)
-        memcached_filter.filter(event)
-      end
-    end
-
-    context 'when the event does not include the value being set' do
-      let(:data) { { "success" => "true" } }
-
-      it 'does not set data on memcached' do
-        expect(cache).to_not receive(:set)
-        memcached_filter.filter(event)
+      after(:each) do
+        memcached_filter.close
       end
 
-      it 'does not invoke `filter_matched`' do
-        expect(memcached_filter).to_not receive(:filter_matched)
+      it "fails #get and tag" do
+        allow(cache).to receive(:multi) { |&b| b.call }
+        allow(cache).to receive(:set).with("success/true/answer", "42")
+        expect(cache).to receive(:get_multi).with(["success/true/answer"]).and_raise(Dalli::RingError.new("foo"))
+        expect(memcached_filter).not_to receive(:filter_matched)
+        memcached_filter.filter(event)
+        expect(event.get("tags")).to eq(["_memcached_failure"])
+      end
+
+      it "fails #set and tag" do
+        allow(cache).to receive(:get_multi).with(["success/true/answer"]).and_return({"success/true/answer" => "42"})
+        allow(cache).to receive(:multi) { |&b| b.call }
+        expect(cache).to receive(:set).with("success/true/answer", "42").and_raise(Dalli::RingError.new("foo"))
+        expect(memcached_filter).not_to receive(:filter_matched)
+        memcached_filter.filter(event)
+        expect(event.get("tags")).to eq(["_memcached_failure"])
+      end
+
+      it "reconnects after a failure" do
+        allow(memcached_filter).to receive(:do_get).and_raise(Dalli::RingError.new("foo"))
+        allow(memcached_filter).to receive(:do_set).and_raise(Dalli::RingError.new("foo"))
+
+        expect(memcached_filter).to receive(:new_connection).once.and_raise("foo")
+        memcached_filter.filter(event)
         memcached_filter.filter(event)
       end
     end

